sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,777 @@
+    local_max_2 = []
+
+    for i in local_maxima_points:
+        if i < s:
+            local_max_1.append(i)
+        if i > s:
+            local_max_2.append(i)
+
+    node_list = []
+    node_list.append((node_1, local_max_1))
+    node_list.append((node_2, local_max_2))
+
+    if reachability_plot[s] < significant_min:
+        node.assign_split_point(-1)
+        # if split_point is not significant, ignore this split and continue
+        _cluster_tree(node, parent_node, local_maxima_points,
+                      reachability_plot, reachability_ordering,
+                      min_cluster_size, maxima_ratio, rejection_ratio,
+                      similarity_threshold, significant_min)
+        return
+
+    # only check a certain ratio of points in the child
+    # nodes formed to the left and right of the maxima
+    # ...should check_ratio be a user settable parameter?
+    check_ratio = .8
+    check_value_1 = int(np.round(check_ratio * len(node_1.points)))
+    check_value_2 = int(np.round(check_ratio * len(node_2.points)))
+    if check_value_2 == 0:
+        check_value_2 = 1
+    avg_reach_value_1 = float(np.average(reachability_plot[(node_1.end -
+                                         check_value_1):node_1.end]))
+    avg_reach_value_2 = float(np.average(reachability_plot[node_2.start:(
+                                node_2.start + check_value_2)]))
+
+    if (float(avg_reach_value_1 / float(reachability_plot[s])) >
+        maxima_ratio or float(avg_reach_value_2 /
+                              float(reachability_plot[s])) > maxima_ratio):
+
+        if float(avg_reach_value_1 /
+                 float(reachability_plot[s])) < rejection_ratio:
+            # reject node 2
+            node_list.remove((node_2, local_max_2))
+        if float(avg_reach_value_2 /
+                 float(reachability_plot[s])) < rejection_ratio:
+            # reject node 1
+            node_list.remove((node_1, local_max_1))
+        if (float(avg_reach_value_1 /
+            float(reachability_plot[s])) >= rejection_ratio and float(
+                avg_reach_value_2 /
+                float(reachability_plot[s])) >= rejection_ratio):
+            node.assign_split_point(-1)
+            # since split_point is not significant,
+            # ignore this split and continue (reject both child nodes)
+            _cluster_tree(node, parent_node, local_maxima_points,
+                          reachability_plot, reachability_ordering,
+                          min_cluster_size, maxima_ratio, rejection_ratio,
+                          similarity_threshold, significant_min)
+            return
+
+    # remove clusters that are too small
+    if (len(node_1.points) < min_cluster_size and
+            node_list.count((node_1, local_max_1)) > 0):
+        # cluster 1 is too small
+        node_list.remove((node_1, local_max_1))
+    if (len(node_2.points) < min_cluster_size and
+            node_list.count((node_2, local_max_1)) > 0):
```

Comment:
I think you're correct about this-- I've changed it.

Checking the correctness of the auto-clustering is difficult, since there isn't a reference implementation of this particular extraction method. The extract_dbscan() method can be tested, since we know what the expected results are when comparing with sklearn's dbscan or another implementation. We can also check the validity of the reachability tree, which is done via comparison to a reference implementation of the OPTICS output on a small test dataset. The automatic cluster extraction is an extension that's extracting via the reachability tree... so we do verify that the input to the method is valid (i.e., the reachability calculations), and we also do verify that the dbscan extraction is valid...

Since there's not another reference implementation that I know of to compare against, I've been gauging it's performance on an extraction test case that uses variable density clusters; it's able to recover the number and extent of those clusters.

Reply:
So you don't have a reference implementation of the auto-extraction, but this is still following a published algorithm, right?? (Just to double-check.)

Yes, I imagine this is hard to test. Possible to construct toy cases that test every path? Possible to design sensible invariants that should be held by it? Perhaps not possible to practically test... but I'm glad I sighted this bug...

Reply:
Yes, we reference 2 published algorithms-- the first is for building the reachability graph and doing the dbscan extraction, the second is the auto-clustering paper. This is the citation in the documentation for the auto-cluster extraction:

“Automatic extraction of clusters from hierarchical clustering representations.” Sander, Jörg, Xuejie Qin, Zhiyong Lu, Nan Niu, and Alex Kovarsky. In Advances in knowledge discovery and data mining, pp. 75-87. Springer Berlin Heidelberg, 2003.